cardano-explorer-node/src/Explorer/Node.hs

```diff
@@ -401,33 +423,57 @@ localTxSubmissionCodec =
+          , recvMsgIntersectNotFound = \  tip -> pure $ go policy Zero currentTip (tipBlockNo tip)
           }
   where
-    clientStIdle :: ClientStIdle blk (Tip blk) m Void
-    clientStIdle = SendMsgRequestNext clientStNext (pure clientStNext)
-
-    clientStNext :: ClientStNext blk (Tip blk) m Void
-    clientStNext =
+    policy = pipelineDecisionLowHighMark 1000 10000
+
+    go :: MkPipelineDecision -> Nat n -> BlockNo -> BlockNo -> ClientPipelinedStIdle n (ByronBlockOrEBB cfg) (Tip blk) m a
+    go mkPipelineDecision n clientTip serverTip =
+      case (n, runPipelineDecision mkPipelineDecision n clientTip serverTip) of
+        (_Zero, (Request, mkPipelineDecision')) ->
+            SendMsgRequestNext clientStNext (pure clientStNext)
+          where
+            clientStNext = mkClientStNext $ \clientBlockNo newServerTip -> go mkPipelineDecision' n clientBlockNo (tipBlockNo newServerTip)
+        (_, (Pipeline, mkPipelineDecision')) ->
+          SendMsgRequestNextPipelined
+            (go mkPipelineDecision' (Succ n) clientTip serverTip)
+        (Succ n', (CollectOrPipeline, mkPipelineDecision')) ->
+          CollectResponse
+            (Just $ SendMsgRequestNextPipelined $ go mkPipelineDecision' (Succ n) clientTip serverTip)
+            (mkClientStNext $ \clientBlockNo newServerTip -> go mkPipelineDecision' n' clientBlockNo (tipBlockNo newServerTip))
+        (Succ n', (Collect, mkPipelineDecision')) ->
+          CollectResponse
+            Nothing
+            (mkClientStNext $ \clientBlockNo newServerTip -> go mkPipelineDecision' n' clientBlockNo (tipBlockNo newServerTip))
+
+    mkClientStNext :: (BlockNo -> Tip blk -> ClientPipelinedStIdle n (ByronBlockOrEBB ByronConfig) (Tip blk) m a)
+                    -> ClientStNext n (ByronBlockOrEBB ByronConfig) (Tip (ByronBlockOrEBB ByronConfig)) m a
+    mkClientStNext finish =
       ClientStNext
-        { recvMsgRollForward = \ blk tip -> ChainSyncClient $ do
-            insertByronBlockOrEBB trce blk (tipBlockNo tip)
-            pure clientStIdle
-        , recvMsgRollBackward = \ point _tip -> ChainSyncClient $ do
-            -- we are requested to roll backward to point 'point', the core
-            -- node's chain's tip is 'tip'.
-            liftIO $ rollbackToPoint trce point
-            pure clientStIdle
+        { recvMsgRollForward = \blk tip -> do
+            liftIO $ do
+              Gauge.set (fromIntegral . unBlockNo $ tipBlockNo tip) $ mNodeHeight metrics
+              newSize <- atomically $ do
+                writeDbActionQueue actionQueue $ DbApplyBlock blk (tipBlockNo tip)
+                lengthDbActionQueue actionQueue
+              Gauge.set (fromIntegral newSize) $ mQueuePostWrite metrics
+            pure $ finish (blockNo blk) tip
+        , recvMsgRollBackward = \point tip -> do
+            newTip <- liftIO $ do
+              atomically $ writeDbActionQueue actionQueue (DbRollBackToPoint point)
+              getCurrentTipBlockNo
```

Comment:
this code still has a small bug that i had left a `TODO` on
`newTip` should be the tip that we just rolled back to, but because of the queue, its not done yet, and `getCurrentTipBlockNo` gets the wrong number

i think the only side-efect is that the pipeline policy may choose to request more blocks, and fall behind, until `recvMsgRollForward` repairs things

Reply:
I agree, but I do not think this is a big enough problem to worry about.